File: nodejs/bin/camera-control.js
#!/usr/bin/env node
/** Command-line interface for Camera Control automation. */

import { mkdir, writeFile } from 'node:fs/promises';
import { parseArgs } from 'node:util';
import { dirname } from 'node:path';
import { randomUUID } from 'node:crypto';
import process from 'node:process';
import readline from 'node:readline';

import {
  ApiError,
  CameraControlClient,
  ConnectionInterrupted,
  ProtocolError,
  TERMINAL_SESSION_STATES,
  decodePhoto,
} from '../src/index.js';

const RESOLUTIONS = ['low', 'medium', 'high', 'veryHigh', 'ultraHigh', 'max'];
const FLASH_MODES = ['off', 'auto', 'always', 'torch'];

const CONNECTION_OPTIONS = {
  ca: { type: 'string' },
  'password-env': { type: 'string', default: 'CAMERA_CONTROL_PASSWORD' },
  timeout: { type: 'string', default: '30' },
  'reconnect-attempts': { type: 'string', default: '5' },
  help: { type: 'boolean', short: 'h', default: false },
};

const CAPTURE_OPTIONS = {
  camera: { type: 'string', default: '1' },
  resolution: { type: 'string', default: 'high' },
  flash: { type: 'string', default: 'off' },
  'jpeg-quality': { type: 'string' },
};

const COMMANDS = {
  status: { ...CONNECTION_OPTIONS, compact: { type: 'boolean', default: false } },
  capture: {
    ...CONNECTION_OPTIONS,
    ...CAPTURE_OPTIONS,
    output: { type: 'string', default: 'photo.jpg' },
    'request-id': { type: 'string' },
  },
  'time-lapse': {
    ...CONNECTION_OPTIONS,
    ...CAPTURE_OPTIONS,
    interval: { type: 'string' },
    count: { type: 'string' },
    duration: { type: 'string' },
    poll: { type: 'string', default: '1' },
  },
};

const USAGE = `Control Camera Control through its verified-TLS WSS API.

usage: camera-control <command> <wss-url> [options]

commands:
  status       print camera, device, capture-session, and reliability status
  capture      capture and save one JPEG through the idempotent job API
  time-lapse   start, poll, and cancel a time-lapse capture session

connection options:
  --ca FILE                 verified Camera Control local CA certificate (required)
  --password-env NAME       environment variable holding the password
                            (default: CAMERA_CONTROL_PASSWORD)
  --timeout SECONDS         request timeout (default: 30)
  --reconnect-attempts N    idempotent retry attempts (default: 5)

capture options (capture, time-lapse):
  --camera N                camera index (default: 1)
  --resolution NAME         ${RESOLUTIONS.join(', ')} (default: high)
  --flash MODE              ${FLASH_MODES.join(', ')} (default: off)
  --jpeg-quality N          1-100

capture options:
  --output FILE             JPEG output path (default: photo.jpg)
  --request-id ID           stable idempotency key; generated when omitted

time-lapse options:
  --interval SECONDS        seconds between photos (required)
  --count N                 stop after this many photos
  --duration SECONDS        cancel after this many seconds
  --poll SECONDS            status poll interval (default: 1)
`;

async function main(argv) {
  const [command, ...rest] = argv;
  if (command === undefined || command === '--help' || command === '-h' || command === 'help') {
    process.stderr.write(USAGE);
    return command === undefined ? 2 : 0;
  }
  const options = COMMANDS[command];
  if (options === undefined) throw new Error(`unknown command '${command}'`);

  const positional = [];
  if (rest.length > 0 && !rest[0].startsWith('-')) positional.push(rest.shift());
  const { values } = parseArgs({ args: rest, options, allowPositionals: false });
  if (values.help) {
    process.stderr.write(USAGE);
    return 0;
  }

  const url = positional[0];
  if (url === undefined) throw new Error('the wss:// URL shown by the app is required');
  if (values.ca === undefined) throw new Error('--ca is required');

  const controls = buildControls(values);
  if (command !== 'status') {
    requireChoice('--resolution', values.resolution, RESOLUTIONS);
    requireChoice('--flash', values.flash, FLASH_MODES);
  }
  const password = await readPassword(values['password-env']);

  const client = new CameraControlClient(url, password, values.ca, {
    timeout: number(values.timeout, '--timeout') * 1000,
    reconnectAttempts: integer(values['reconnect-attempts'], '--reconnect-attempts'),
  });
  await client.connect();
  try {
    if (command === 'status') return await runStatus(client, values);
    if (command === 'capture') return await runCapture(client, values, controls);
    return await runTimeLapse(client, values, controls);
  } finally {
    client.close();
  }
}

async function runStatus(client, values) {
  const status = await client.status();
  process.stdout.write(`${stableJson(status, values.compact)}\n`);
  return 0;
}

async function runCapture(client, values, controls) {
  const requestId = values['request-id'] ?? `node-cli-${randomUUID()}`;
  const job = await client.captureJob({
    camera: integer(values.camera, '--camera'),
    resolution: values.resolution,
    flash: values.flash,
    controls,
    requestId,
  });
  if (job.state !== 'completed') {
    throw new ProtocolError(`Capture ended as ${job.state}: ${job.error?.message ?? ''}`);
  }
  const photo = job.result?.photo;
  if (photo === undefined) throw new ProtocolError('Completed capture job has no photo');
  const jpeg = decodePhoto(photo);
  const output = values.output;
  const parent = dirname(output);
  if (parent !== '.') await mkdir(parent, { recursive: true });
  await writeFile(output, jpeg);
  process.stdout.write(`Saved ${jpeg.length} bytes to ${output} (requestId=${requestId})\n`);
  return 0;
}

async function runTimeLapse(client, values, controls) {
  const interval = number(values.interval, '--interval');
  if (!(interval >= 1 && interval <= 86400)) {
    throw new Error('--interval must be between 1 and 86400 seconds');
  }
  const count = values.count === undefined ? undefined : integer(values.count, '--count');
  if (count !== undefined && !(count >= 2 && count <= 1000)) {
    throw new Error('--count must be between 2 and 1000');
  }
  const duration = values.duration === undefined ? undefined : number(values.duration, '--duration');
  if (duration !== undefined && !(duration > 0)) throw new Error('--duration must be positive');
  const poll = number(values.poll, '--poll');
  if (!(poll > 0)) throw new Error('--poll must be positive');

  let session = await client.startTimeLapse({
    intervalSeconds: interval,
    count,
    camera: integer(values.camera, '--camera'),
    resolution: values.resolution,
    flash: values.flash,
    controls,
  });
  const sessionId = String(session.id);
  process.stdout.write(`Started ${sessionId}; press Ctrl+C to stop.\n`);

  let stopRequested = false;
  const onInterrupt = () => {
    process.stderr.write('\nCancellation requested…\n');
    stopRequested = true;
  };
  process.on('SIGINT', onInterrupt);
  const deadline = duration === undefined ? undefined : Date.now() + duration * 1000;
  let lastSignature;
  try {
    while (!TERMINAL_SESSION_STATES.includes(session.state)) {
      const signature = `${session.state}/${session.capturedCount}`;
      if (signature !== lastSignature) {
        process.stdout.write(
          `${session.state}: ${session.capturedCount}/${session.targetCount ?? '∞'}\n`,
        );
        lastSignature = signature;
      }
      if (stopRequested || (deadline !== undefined && Date.now() >= deadline)) {
        session = await client.cancelCaptureSession(sessionId);
        break;
      }
      await sleep(poll * 1000);
      if (stopRequested) {
        session = await client.cancelCaptureSession(sessionId);
        break;
      }
      session = await client.captureSessionStatus(sessionId);
    }
  } finally {
    process.off('SIGINT', onInterrupt);
  }

  process.stdout.write(`${stableJson(session, false)}\n`);
  return ['completed', 'cancelled'].includes(session.state) ? 0 : 1;
}

function buildControls(values) {
  const quality = values['jpeg-quality'];
  if (quality === undefined) return {};
  const parsed = integer(quality, '--jpeg-quality');
  if (parsed < 1 || parsed > 100) throw new Error('--jpeg-quality must be between 1 and 100');
  return { jpegQuality: parsed };
}

function requireChoice(name, value, allowed) {
  if (!allowed.includes(value)) throw new Error(`${name} must be one of ${allowed.join(', ')}`);
}

function number(value, name) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) throw new Error(`${name} must be a number`);
  return parsed;
}

function integer(value, name) {
  const parsed = Number(value);
  if (!Number.isInteger(parsed)) throw new Error(`${name} must be an integer`);
  return parsed;
}

/** Serialize with sorted keys so that output diffs stay readable. */
function stableJson(value, compact) {
  const sortedKeys = new Set();
  const collect = (node) => {
    if (node !== null && typeof node === 'object' && !Array.isArray(node)) {
      for (const key of Object.keys(node).sort()) {
        sortedKeys.add(key);
        collect(node[key]);
      }
    } else if (Array.isArray(node)) {
      node.forEach(collect);
    }
  };
  collect(value);
  return JSON.stringify(value, [...sortedKeys], compact ? undefined : 2);
}

function sleep(milliseconds) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

/**
 * Prefer the environment variable so that the password never appears in
 * process arguments, and prompt without echo otherwise.
 */
async function readPassword(variable) {
  const fromEnvironment = process.env[variable];
  if (fromEnvironment !== undefined) return fromEnvironment;
  const rl = readline.createInterface({ input: process.stdin, output: process.stderr, terminal: true });
  const question = new Promise((resolve) => rl.question('Camera Control password: ', resolve));
  rl.output.write = () => {};
  const password = await question;
  rl.close();
  process.stderr.write('\n');
  return password;
}

try {
  process.exitCode = await main(process.argv.slice(2));
} catch (error) {
  const known =
    error instanceof ApiError ||
    error instanceof ConnectionInterrupted ||
    error instanceof ProtocolError ||
    error instanceof Error;
  process.stderr.write(`camera-control: ${known ? error.message : String(error)}\n`);
  process.exitCode = 1;
}

File: rust/src/bin/camera_control.rs
//! Command-line interface for Camera Control automation.

use std::io::{BufRead, Write};
use std::path::PathBuf;
use std::process::{Command as ProcessCommand, Stdio};
use std::time::{Duration, Instant};

use camera_control::{
    decode_photo, CaptureOptions, Client, Error, Options, Result, TERMINAL_SESSION_STATES,
};
use clap::{Args, Parser, Subcommand, ValueEnum};
use serde_json::{Map, Value};

#[derive(Parser)]
#[command(
    name = "camera-control",
    about = "Control Camera Control through its verified-TLS WSS API.",
    version
)]
struct Cli {
    #[command(subcommand)]
    command: CommandKind,
}

#[derive(Subcommand)]
enum CommandKind {
    /// Print camera, device, capture-session, and reliability status.
    Status {
        #[command(flatten)]
        connection: Connection,
        /// Print compact JSON.
        #[arg(long)]
        compact: bool,
    },
    /// Capture and save one JPEG through the idempotent job API.
    Capture {
        #[command(flatten)]
        connection: Connection,
        #[command(flatten)]
        capture: CaptureArgs,
        /// JPEG output path.
        #[arg(long, default_value = "photo.jpg")]
        output: PathBuf,
        /// Stable idempotency key; generated when omitted.
        #[arg(long = "request-id")]
        request_id: Option<String>,
    },
    /// Start, poll, and cancel a time-lapse capture session.
    TimeLapse {
        #[command(flatten)]
        connection: Connection,
        #[command(flatten)]
        capture: CaptureArgs,
        /// Seconds between photos.
        #[arg(long, value_parser = interval_seconds)]
        interval: f64,
        /// Stop after this many photos.
        #[arg(long, value_parser = clap::value_parser!(i64).range(2..=1000))]
        count: Option<i64>,
        /// Cancel after this many seconds.
        #[arg(long, value_parser = positive_seconds)]
        duration: Option<f64>,
        /// Status poll interval in seconds.
        #[arg(long, default_value_t = 1.0, value_parser = positive_seconds)]
        poll: f64,
    },
}

#[derive(Args)]
struct Connection {
    /// The wss:// URL shown by the app.
    url: String,
    /// Verified Camera Control local CA certificate.
    #[arg(long)]
    ca: PathBuf,
    /// Environment variable containing the password.
    #[arg(long = "password-env", default_value = "CAMERA_CONTROL_PASSWORD")]
    password_env: String,
    /// Request timeout in seconds.
    #[arg(long, default_value_t = 30.0, value_parser = positive_seconds)]
    timeout: f64,
    /// Idempotent retry attempts.
    #[arg(long = "reconnect-attempts", default_value_t = 5)]
    reconnect_attempts: u32,
}

#[derive(Args)]
struct CaptureArgs {
    /// Camera index.
    #[arg(long, default_value_t = 1)]
    camera: i64,
    #[arg(long, value_enum, default_value_t = Resolution::High)]
    resolution: Resolution,
    #[arg(long, value_enum, default_value_t = Flash::Off)]
    flash: Flash,
    /// JPEG quality between 1 and 100.
    #[arg(long = "jpeg-quality", value_parser = clap::value_parser!(i64).range(1..=100))]
    jpeg_quality: Option<i64>,
}

#[derive(Clone, Copy, ValueEnum)]
enum Resolution {
    Low,
    Medium,
    High,
    #[value(name = "veryHigh")]
    VeryHigh,
    #[value(name = "ultraHigh")]
    UltraHigh,
    Max,
}

#[derive(Clone, Copy, ValueEnum)]
enum Flash {
    Off,
    Auto,
    Always,
    Torch,
}

impl CaptureArgs {
    fn to_options(&self, request_id: Option<String>) -> CaptureOptions {
        let mut controls = Map::new();
        if let Some(quality) = self.jpeg_quality {
            controls.insert("jpegQuality".into(), Value::from(quality));
        }
        CaptureOptions {
            camera: self.camera,
            resolution: self
                .resolution
                .to_possible_value()
                .expect("every resolution has a value")
                .get_name()
                .to_string(),
            flash: self
                .flash
                .to_possible_value()
                .expect("every flash mode has a value")
                .get_name()
                .to_string(),
            controls,
            request_id,
        }
    }
}

#[tokio::main]
async fn main() -> std::process::ExitCode {
    match run(Cli::parse()).await {
        Ok(code) => code,
        Err(error) => {
            eprintln!("camera-control: {error}");
            std::process::ExitCode::FAILURE
        }
    }
}

async fn run(cli: Cli) -> Result<std::process::ExitCode> {
    let connection = match &cli.command {
        CommandKind::Status { connection, .. }
        | CommandKind::Capture { connection, .. }
        | CommandKind::TimeLapse { connection, .. } => connection,
    };
    let password = read_password(&connection.password_env)?;
    let mut client = Client::with_ca_file(
        connection.url.clone(),
        password,
        &connection.ca,
        Options {
            timeout: Duration::from_secs_f64(connection.timeout),
            reconnect_attempts: connection.reconnect_attempts,
            ..Options::default()
        },
    )
    .await?;
    client.connect().await?;

    let code = match &cli.command {
        CommandKind::Status { compact, .. } => {
            let status = client.status().await?;
            println!("{}", render(&status, *compact));
            std::process::ExitCode::SUCCESS
        }
        CommandKind::Capture {
            capture,
            output,
            request_id,
            ..
        } => run_capture(&mut client, capture, output, request_id.clone()).await?,
        CommandKind::TimeLapse {
            capture,
            interval,
            count,
            duration,
            poll,
            ..
        } => run_time_lapse(&mut client, capture, *interval, *count, *duration, *poll).await?,
    };
    client.close().await;
    Ok(code)
}

async fn run_capture(
    client: &mut Client,
    capture: &CaptureArgs,
    output: &PathBuf,
    request_id: Option<String>,
) -> Result<std::process::ExitCode> {
    let request_id = request_id.unwrap_or_else(|| format!("rust-cli-{}", uuid::Uuid::new_v4()));
    let options = capture.to_options(Some(request_id.clone()));
    let job = client.capture_job(&options).await?;
    if job.get("state").and_then(Value::as_str) != Some("completed") {
        let message = job
            .get("error")
            .and_then(|error| error.get("message"))
            .and_then(Value::as_str)
            .unwrap_or_default();
        return Err(Error::Protocol(format!(
            "capture ended as {}: {message}",
            job.get("state").unwrap_or(&Value::Null)
        )));
    }
    let photo = job
        .get("result")
        .and_then(|result| result.get("photo"))
        .ok_or_else(|| Error::Protocol("completed capture job has no photo".into()))?;
    let jpeg = decode_photo(photo)?;
    if let Some(parent) = output.parent().filter(|parent| !parent.as_os_str().is_empty()) {
        tokio::fs::create_dir_all(parent).await?;
    }
    tokio::fs::write(output, &jpeg).await?;
    println!(
        "Saved {} bytes to {} (requestId={request_id})",
        jpeg.len(),
        output.display()
    );
    Ok(std::process::ExitCode::SUCCESS)
}

async fn run_time_lapse(
    client: &mut Client,
    capture: &CaptureArgs,
    interval: f64,
    count: Option<i64>,
    duration: Option<f64>,
    poll: f64,
) -> Result<std::process::ExitCode> {
    let options = capture.to_options(None);
    let mut session = client.start_time_lapse(interval, count, &options).await?;
    let session_id = session
        .get("id")
        .and_then(Value::as_str)
        .ok_or_else(|| Error::Protocol("the session has no id".into()))?
        .to_string();
    println!("Started {session_id}; press Ctrl+C to stop.");

    let deadline = duration.map(|seconds| Instant::now() + Duration::from_secs_f64(seconds));
    let mut last_signature = String::new();
    let mut cancel = false;
    while !is_terminal(&session) {
        let signature = format!(
            "{}/{}",
            field(&session, "state"),
            field(&session, "capturedCount")
        );
        if signature != last_signature {
            let target = match session.get("targetCount") {
                Some(Value::Null) | None => "∞".to_string(),
                Some(value) => value.to_string(),
            };
            println!(
                "{}: {}/{target}",
                field(&session, "state"),
                field(&session, "capturedCount")
            );
            last_signature = signature;
        }
        if deadline.is_some_and(|deadline| Instant::now() >= deadline) {
            cancel = true;
            break;
        }
        tokio::select! {
            _ = tokio::time::sleep(Duration::from_secs_f64(poll)) => {}
            _ = tokio::signal::ctrl_c() => {
                eprintln!("\nCancellation requested…");
                cancel = true;
            }
        }
        if cancel {
            break;
        }
        session = client.capture_session_status(&session_id).await?;
    }
    if cancel {
        session = client.cancel_capture_session(&session_id).await?;
    }

    println!("{}", render(&session, false));
    Ok(match session.get("state").and_then(Value::as_str) {
        Some("completed" | "cancelled") => std::process::ExitCode::SUCCESS,
        _ => std::process::ExitCode::FAILURE,
    })
}

fn is_terminal(session: &Value) -> bool {
    session
        .get("state")
        .and_then(Value::as_str)
        .is_some_and(|state| TERMINAL_SESSION_STATES.contains(&state))
}

fn field(value: &Value, key: &str) -> String {
    match value.get(key) {
        Some(Value::String(text)) => text.clone(),
        Some(other) => other.to_string(),
        None => "null".into(),
    }
}

/// serde_json keeps object keys sorted only with the `preserve_order` feature
/// disabled, which is the default, so plain serialization is already stable.
fn render(value: &Value, compact: bool) -> String {
    if compact {
        value.to_string()
    } else {
        serde_json::to_string_pretty(value).unwrap_or_else(|_| value.to_string())
    }
}

fn interval_seconds(raw: &str) -> std::result::Result<f64, String> {
    let seconds: f64 = raw.parse().map_err(|_| "expected a number".to_string())?;
    if !(1.0..=86_400.0).contains(&seconds) {
        return Err("must be between 1 and 86400 seconds".into());
    }
    Ok(seconds)
}

fn positive_seconds(raw: &str) -> std::result::Result<f64, String> {
    let seconds: f64 = raw.parse().map_err(|_| "expected a number".to_string())?;
    if seconds <= 0.0 {
        return Err("must be positive".into());
    }
    Ok(seconds)
}

/// Prefer the environment variable so that the password never appears in
/// process arguments, and prompt without echo otherwise.
fn read_password(variable: &str) -> Result<String> {
    if let Ok(password) = std::env::var(variable) {
        return Ok(password);
    }
    eprint!("Camera Control password: ");
    std::io::stderr().flush().ok();
    let echo_disabled = set_terminal_echo(false);
    let mut password = String::new();
    let read = std::io::stdin().lock().read_line(&mut password);
    if echo_disabled {
        set_terminal_echo(true);
    }
    eprintln!();
    read.map_err(|error| Error::Config(error.to_string()))?;
    Ok(password.trim_end_matches(['\r', '\n']).to_string())
}

fn set_terminal_echo(enabled: bool) -> bool {
    ProcessCommand::new("stty")
        .arg(if enabled { "echo" } else { "-echo" })
        .stdin(Stdio::inherit())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|status| status.success())
        .unwrap_or(false)
}
